ec/ecresource/deploymentresource/apm_expanders.go

```diff
@@ -162,3 +160,28 @@ func expandApmConfig(raw interface{}) *models.ApmConfiguration {
 
 	return nil
 }
+
+func defaultApmTopology(topology []*models.ApmTopologyElement) []*models.ApmTopologyElement {
+	for _, t := range topology {
+		if *t.Size.Value > defaultApmSize {
+			t.Size.Value = ec.Int32(defaultApmSize)
+		}
+		if t.ZoneCount > 1 {
```

Comment:
I don't really understand these checks? Why are we checking if size and zone count are larger than the defaults to set the defaults?

Reply:
Yeah sorry, I'll add a comment on the code to solve that.

Basically, there's a few scenarios that I'm trying to cover here in the APM resource, at the moment, the deployment template is as follows: 

```javascript
{
    "ref_id": "apm-ref-id",
    "elasticsearch_cluster_ref_id": "es-ref-id",
    "region": "us-east-1",
    "plan": {
        "cluster_topology": [
            {
                "instance_configuration_id": "aws.apm.r5d",
                "size": {
                    "value": 512,
                    "resource": "memory"
                },
                "zone_count": 1
            }
        ],
        "apm": {}
    }
}
```

but this could change in an upcoming deployment template, to have a consistent behavior across the board, what I've done is:

-  if the size is > than what we say is the default in the `memory_per_node`, then an override with that default is set.
- if the zone_count is > 1 then, we apply the default 1.

Apm and Kibana have the same topology pre say, the only thing that changes is the `instance_configuration_id`, but the template for Enterprise Search is sets the `size.value: 0`, which then we need to check in that function as we do.

Elasticsearch is slightly different, where the ES deployment template contains all the available templates for each of the instance configurations enabling only a some of the blueprinted ones:

```javascript
{
    "ref_id": "es-ref-id",
    "region": "us-east-1",
    "plan": {
        "cluster_topology": [
            {
                "node_type": {
                    "master": false,
                    "data": false,
                    "ingest": true
                },
                "zone_count": 2,
                "instance_configuration_id": "aws.coordinating.m5d",
                "size": {
                    "value": 0,
                    "resource": "memory"
                }
            },
            {
                "node_type": {
                    "master": true,
                    "data": true,
                    "ingest": true
                },
                "zone_count": 2,
                "instance_configuration_id": "aws.data.highio.i3",
                "size": {
                    "value": 8192,
                    "resource": "memory"
                }
            },
            {
                "node_type": {
                    "master": true,
                    "data": false,
                    "ingest": false
                },
                "zone_count": 3,
                "instance_configuration_id": "aws.master.r5d",
                "size": {
                    "value": 0,
                    "resource": "memory"
                }
            },
            {
                "node_type": {
                    "master": false,
                    "data": false,
                    "ingest": false,
                    "ml": true
                },
                "zone_count": 1,
                "instance_configuration_id": "aws.ml.m5d",
                "size": {
                    "value": 0,
                    "resource": "memory"
                }
            }
        ],
        "elasticsearch": {}
    },
    "settings": {
        "dedicated_masters_threshold": 6
    }
}
```

In the case above ^, the instance configuration which is enabled by default is `"aws.data.highio.i3", which has a size of `8192` and `zone_count` set to `2`. This is not what we advertise as defaults in the terraform provider and thus, to provide a consistent experience, is why we check the ZoneCount and the size.

I've answered here all of the other comments, I'm going to mark the others as resolved and we can keep this conversation here.

Reply:
Thanks for the explanation 🙇‍♀️ 

> In the case above ^, the instance configuration which is enabled by default is "aws.data.highio.i3", which has a size of 8192 and zone_count set to 2. This is not what we advertise as defaults in the terraform provider

Should we consider having those defaults on the provider as well? I'm thinking those defaults are on the template for a reason, maybe we shouldn't change them? 🤔

Reply:
I was thinking about it... I think we should move to it, my only doubt are the size for the trials in ESS, I'm not sure they're allowed to spin 2x 8Gb, but I'll confirm this today.